### PR TITLE
Improve tests for `index_fill_`

### LINKED ISSUE #24714: Migrate `index_fill_` from the TH to Aten (CPU)
Porting TH operators is essential for code simplicity and performance reasons.

Porting guides and Q&A are available in umbrella issue: #24507

Feel free to add @VitalyFedyunin as a reviewer to get a prioritized review.

cc @ezyang @gchanan @zou3519 @bdhirsh @jbschlosser

### PR DESCRIPTION
**Pull request by kshitij12345**
~~Fixes #24714~~

Improves test for `index_fill_`

### CODE REVIEW

File: test/test_torch.py [hunk 1]
```diff
@@ -4303,14 +4303,72 @@ def test_func(slf, device, call_type):
         test_func(self, device, 'method inplace')
 
     def test_index_fill(self, device):
+        def _verify_against_expected(x, idx, val, expected):
+            x.index_fill_(1, idx, val)
+            self.assertEqual(x, expected)
+
         for dt in torch.testing.get_all_dtypes():
```
Comment (mruberry) [inline test/test_torch.py:4310]: Use the `@dtypes` decorator instead of enumerating dtypes in the test

File: test/test_torch.py [hunk 2]
```diff
@@ -4303,14 +4303,72 @@ def test_func(slf, device, call_type):
...
             if dt == torch.half or dt == torch.bfloat16 or dt.is_complex:
                 continue
 
-            x = torch.tensor([[1, 2], [4, 5]], dtype=dt, device=device)
+            val = 0
```
Comment (mruberry) [inline test/test_torch.py:4314]: Add comments describing what's actually being tested

File: test/test_torch.py [hunk 3]
```diff
@@ -4303,14 +4303,72 @@ def test_func(slf, device, call_type):
...
+
+            x_non_contig = torch.tensor([[1, 2, 3], [4, 5, 6]], dtype=dt, device=device)[:, :2]
+            self.assertTrue(not x_non_contig.is_contiguous())
+            _verify_against_expected(x_non_contig, index, val, expected_contig_idx)
+
+            x_non_contig = torch.tensor([[1, 2, 3, 4], [4, 5, 6, 7]], dtype=dt, device=device)[:, :3]
+            self.assertTrue(not x_non_contig.is_contiguous())
+            _verify_against_expected(x_non_contig, index_non_contig, val, expected_non_contig_idx)
+
+            # Verify `index` should be a LongTensor
```
Comment (mruberry) [inline test/test_torch.py:4336]: Make this an ErrorInput in the OpInfo for `index_fill_`

File: test/test_torch.py [hunk 4]
```diff
@@ -4303,14 +4303,72 @@ def test_func(slf, device, call_type):
...
+            for idx in range(3, 30, 2):
+                expected_noncontig[idx] = val
+            x_noncontig.index_fill_(0, index, val)
+            self.assertEqual(x_noncontig, torch.tensor(expected_noncontig, dtype=dt, device=device))
+
+            x_zero_dim = torch.tensor(0, dtype=dt, device=device)
+            x_zero_dim.index_fill_(0, torch.tensor([0], device=device), 3)
+            self.assertEqual(x_zero_dim, torch.tensor(3, dtype=dt, device=device))
+
+            # Verify IndexError is raised
```
Comment (mruberry) [inline test/test_torch.py:4363]: Make this an ErrorInput, too

File: test/test_torch.py [hunk 5]
```diff
@@ -4303,14 +4303,72 @@ def test_func(slf, device, call_type):
         test_func(self, device, 'method inplace')
 
     def test_index_fill(self, device):
+        def _verify_against_expected(x, idx, val, expected):
```
Comment (mruberry) [inline test/test_torch.py:4306]: How hard would it be to implement `torch.Tensor.index_fill` using NumPy operators? That would let us add a reference to the `index_fill` OpInfo and this test could just validate against the NumPy implementation, which would be simpler and more readable.

**Comment by kshitij12345** [comment]
I think the OpInfo test which check with non-contiguous input are already checking these tests.

**Pull request closed without merging**